### PR TITLE
Fix for discord.User parsing 

### LINKED ISSUE #367: discord.User is not a valid input type for discord.commands.Option()
### Summary

 Statement from @Lulalaby:
> Libs should be able to parse Option(discord.User) and make it Castable as member within the context of the current guild, If applicable.

### Reproduction Steps

- Pass a `discord.User` to the type argument in a `discord.commands.Option()`.

### Minimal Reproducible Code

```python
import discord
...
discord.commands.Option(discord.User)
```

### Expected Results

Get a user object similar to the `discord.Member` class.

### Actual Results

```
Traceback (most recent call last):
  File "C:\Dev\Projects\bot.py", line 75, in <module>
    discord.commands.Option(discord.User)
  File "C:\Users\mendo\AppData\Local\Programs\Python\Python310\lib\site-packages\discord\commands\commands.py", line 564, in __init__
    _type = SlashCommandOptionType.from_datatype(input_type)
  File "C:\Users\mendo\AppData\Local\Programs\Python\Python310\lib\site-packages\discord\enums.py", line 659, in from_datatype
    raise TypeError(f'Invalid class {datatype} used as an input type for an Option')
TypeError: Invalid class <class 'discord.user.User'> used as an input type for an Option
```

### Intents

Im not using intents.

### System Information

```
- Python v3.10.0-final
- py-cord v2.0.0-alpha
    - py-cord pkg_resources: v2.0.0a4214+g5a23a46
- aiohttp v3.7.4.post0
- system info: Windows 10 10.0.22000
```
(Actually, Windows 11 10.0.22000)

### Checklist

- [X] I have searched the open issues for duplicates.
- [X] I have shown the entire traceback, if possible.
- [X] I have removed my token from display, if visible.

### Additional Context

I was trying to make a Slash command with an `Option(discord.Member)` argument and it worked perfectly, except when someone called the command from the bot DM's. 

Since it is a DM and there is no guild, the `Option()` class can not use the `get_member()` method from a NoneType. I tried to implement the @Lulalaby idea of using a `discord.User` instead and if the slash command was called from a guild, then get the discord.Member trough the user ID.

### PR DESCRIPTION
**Pull request by Apocryphon-X**
## Summary

Closes #367. 
Fix for parsing of `discord.Member` / `discord.User` when Slash Command is invoked from a DM. 

## Checklist

- [x] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [x] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting, examples, ...)

### CODE REVIEW

File: discord/commands/commands.py [hunk 1]
```diff
@@ -635,14 +640,19 @@ def __init__(
...
             raise TypeError(f"Expected {minmax_typehint} for max_value, got \"{type(self.max_value).__name__}\"")
 
         self.autocomplete = kwargs.pop("autocomplete", None)
-
+        if (
+            self.autocomplete and
+            not asyncio.iscoroutinefunction(self.autocomplete)
+        ):
+            raise TypeError("Autocomplete callback must be a coroutine.")
+
```
Comment (BobDotCom) [inline discord/commands/commands.py:655]: sync autocomplete callbacks are allowed now

**Review by BobDotCom — CHANGES_REQUESTED** [review]

File: discord/commands/commands.py (hunk 1 above)
Comment (Lulalaby) [reply discord/commands/commands.py]: oops, my fault, sry

**Review by Dorukyum — CHANGES_REQUESTED** [review]
Removing `ctx.` prefixes on lines 538 and 543 creates `NameError`s.

**Comment by Lulalaby** [comment]
@Dorukyum

**Review by Middledot — APPROVED** [review]

**Review by RPSMain — APPROVED** [review]

**Comment by Dorukyum** [comment]
Has this been extensively tested? Just to make sure.

**Review by Dorukyum — APPROVED** [review]

**Pull request merged**